File: python/markdown.py
import re


def parse(markdown):
    s = markdown
    for i in range(6, 0, -1):
        s = re.sub(rf'^{"#" * i} (.*$)', rf'<h{i}>\1</h{i}>', s, flags=re.M)
    s = re.sub(r'__([^\n]+?)__', r'<strong>\1</strong>', s)
    s = re.sub(r'_([^\n]+?)_', r'<em>\1</em>', s)
    s = re.sub(r'^\* (.*$)', r'<li>\1</li>', s, flags=re.M)
    s = re.sub(r'(<li>.*</li>)', r'<ul>\1</ul>', s, flags=re.S)
    s = re.sub(r'^(?!<[hlu])(.*$)', r'<p>\1</p>', s, flags=re.M)
    return s.replace('\n', '')

File: go/markdown.go
package markdown

import (
	"fmt"
	"regexp"
	"strings"
)

// Render translates markdown to HTML
func Render(m string) string {
	for i := 6; i > 0; i-- {
		pattern := fmt.Sprintf(`(?m)^%s (.*)$`, strings.Repeat("#", i))
		replace := fmt.Sprintf("<h%d>$1</h%d>", i, i)
		re := regexp.MustCompile(pattern)
		m = re.ReplaceAllString(m, replace)
	}

	m = regexp.MustCompile(`__(.*)__`).ReplaceAllString(m, "<strong>$1</strong>")
	m = regexp.MustCompile(`_(.*)_`).ReplaceAllString(m, "<em>$1</em>")
	m = regexp.MustCompile(`(?m)^\* (.*)$`).ReplaceAllString(m, "<li>$1</li>")
	m = regexp.MustCompile(`(?s)(<li>.*</li>)`).ReplaceAllString(m, "<ul>$1</ul>")

	s := strings.Builder{}
	r := regexp.MustCompile(`^<[hlu]`)
	for _, line := range strings.Split(m, "\n") {
		if r.MatchString(line) {
			s.WriteString(line)
		} else {
			s.WriteString("<p>" + line + "</p>")
		}
	}
	return s.String()
}
